src/main/java/org/tugraz/sysds/utils/Explain.java

```diff
@@ -547,15 +573,57 @@ else if( hop.requiresReblock() )
 			else if( hop.requiresCheckpoint() )
 				sb.append(" [chkpt]");
 		}
-		
+
 		//exec type
 		if (hop.getExecType() != null)
 			sb.append(", " + hop.getExecType());
-		
+
 		sb.append('\n');
-		
+
 		hop.setVisited();
-		
+
+		return sb.toString();
+	}
+
+	/**
+	 * Do a post-order traverse through the Lineage Item DAG and explain each Hop
+	 *
+	 * @param li lineage item
+	 * @param level offset
+	 * @return string explanation of Lineage Item DAG
+	 */
+	private static String explainLineageItem(LineageItem li, int level) {
+		if( li.isVisited())
+			return "";
+
+		StringBuilder sb = new StringBuilder();
+		String offset = createOffset(level);
+
+		for( LineageItem input : li.getAncestors() )
```

Comment:
maybe rename Ancestor to Input to avoid confusion as it depends the subjective perspective (sometimes the parent in a data flow graph is the consumer as these graphs are drawn bottom-up).

Reply:
Should I rename _children into _outputs?

Reply:
yes, that sounds like a good idea.